test/versioned/fastify/add-hook.tap.js

```diff
@@ -33,51 +48,97 @@ tap.test(function (t) {
 
-  t.test('should wrap all lifecycle hooks', async function (t) {
+  t.test('non-error hooks', async function nonErrorHookTest(t) {
+    // setup fastify route
+    fastify.get('/', function routeHandler(_, reply) {
+      t.comment('route handler called')
+      reply.send({ hello: 'world' })
+    })
+
     // setup hooks
-    const ok = REQUEST_HOOKS.reduce((all, hookName) => {
+    const ok = ALL_HOOKS.reduce((all, hookName) => {
+      t.comment(`registering ${hookName}`)
       all[hookName] = false
       return all
     }, {})
-    for (const hookName of REQUEST_HOOKS) {
-      fastify.addHook(hookName, function testHook(req, res, next) {
+
+    ALL_HOOKS.forEach((hookName) => {
+      fastify.addHook(hookName, function testHook(...args) {
+        t.comment(`${hookName} called`)
+        // lifecycle signatures vary between the events
+        // the last arg is always the next function though
+        const next = args[args.length - 1]
         ok[hookName] = true
         next()
       })
-    }
-    // setup transaction monitor
-    const promise = new Promise((resolve) => {
-      agent.on('transactionFinished', (transaction) => {
-        t.equal('WebFrameworkUri/Fastify/GET//', transaction.getName(), `transaction name matched`)
-        metrics.assertSegments(transaction.trace.root, [
-          'WebTransaction/WebFrameworkUri/Fastify/GET//',
-          REQUEST_HOOKS.map((hookName) => {
-            return `Nodejs/Middleware/Fastify/${hookName}/testHook`
-          })
-        ])
-        resolve()
-      })
     })
-    // start and ping the server
+
+    agent.on('transactionFinished', (transaction) => {
+      t.equal('WebFrameworkUri/Fastify/GET//', transaction.getName(), `transaction name matched`)
+      // all the hooks are siblings of the route handler
+      // except the AFTER_HANDLER_HOOKS which are children of the route handler
+      metrics.assertSegments(transaction.trace.root, [
+        'WebTransaction/WebFrameworkUri/Fastify/GET//',
+        [
+          ...getSegmentNames(REQUEST_HOOKS),
+          'Nodejs/Middleware/Fastify/routeHandler',
+          getSegmentNames(AFTER_HANDLER_HOOKS)
+        ]
+      ])
+    })
+
     await fastify.listen(0)
     const { port } = fastify.server.address()
-    await requestClient.get(`http://127.0.0.1:${port}/`)
-    await promise
-    // validate
+    const result = await getAsync(`http://127.0.0.1:${port}/`)
+    t.equal(result.statusCode, 200)
+    t.equal(result.body, JSON.stringify({ hello: 'world' }))
+
+    // verify every hook was called after response
     for (const [hookName, isOk] of Object.entries(ok)) {
       t.equal(isOk, true, `${hookName} captured`)
     }
     t.end()
   })
+
+  t.test('error hook', async function errorHookTest(t) {
+    // setup fastify route
+    fastify.get('/', function routeHandler() {
+      throw new Error('sup')
+    })
+
+    const hookName = 'onError'
+    let ok = false
+
+    fastify.addHook(hookName, (req, reply, next) => {
```

Comment:
it looks like the signature is diff https://github.com/fastify/fastify/blob/main/docs/Hooks.md#manage-errors-from-a-hook, it has 4 args and next is last